SPIKE-NOTES.md

```diff
@@ -0,0 +1,107 @@
+# Spike notes
+
+In this spike pull request, I am showing how we can allow LB3 developers to
+expose their legacy applications in an LB4 project and thus allow them to run
+both old LB3 and new LB4 endpoints in the same process.
+
+The solution has two major parts:
+
+- `app.mountExpressRouter(basePath, router, spec)`
+- `Lb3AppBooter`
+
+## mountExpressRouter
+
+```ts
+mountExpressRouter(
+  basePath: string,
+  router: ExpressRequestHandler,
+  spec?: RouterSpec,
+): void;
+```
+
+This new API is building on top of StaticAssetsRoute and allows LB4 app
+developers to add arbitrary set of Express routes and provide OpenAPI spec.
+
+The new request-handling pipeline has the following steps now:
+
+1. Request-preprocessing middleware like CORS, this is not customizable yet.
+2. Native LoopBack 4 routes (controller methods or route handlers).
+3. External Express routers (if the request was not handled yet).
+4. Static assets (if the request was not handled yet).
+
+See
+[external-express-routes.ts](./packages/rest/src/router/external-express-routes.ts)
+for most of the implementation details.
+
+## Lb3AppBooter
+
+This booter is responsible for setting up LoopBack 3 application, obtaining
+OpenAPI spec describing app's REST API and mounting the app on LB4.
+
+I am proposing to put this booter into a standalone npm package to allow us to
```

Comment:
> I am proposing to put this booter into a standalone npm package ...

Loose coupling FTW. Getting rid of it from the framework (in future) will also be simpler and neater.